### PR TITLE
executor: detach session tracker for each query

### LINKED ISSUE #17420: Detach memory tracker for insert/update
## Development Task

Currently, for the session memory tracker, it won't detach from the global memory tracker for `insert` and `update` which would cause the inaccurate memory usage statistical information. We need to detach the memory tracker in these cases.

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve?
Currently, we detach session tracker from global tracker in `CloseRecordSet` function. If the query have no `RecordSet` returned (eg: `insert` / `update`), the session tracker won't be detached from global tracker which would cause the inaccurate memory usage statistical information.

Issue Number: close #17420 

### What is changed and how it works?

How it Works:

Detach session tracker for each query when the query ended.

### Release note
* No release note

### CODE REVIEW

**Comment by Yisaer** [comment]
@qw4990 PTAL when you are free.

File: executor/memory_test.go [hunk 1]
```diff
@@ -114,3 +122,17 @@ func (s *testMemoryLeak) memDiff(m1, m2 uint64) uint64 {
...
+	tk.MustExec("use test")
+	tk.MustExec("create table t (id int)")
+	tk.MustExec("insert t (id) values (1)")
+	tk.MustExec("insert t (id) values (2)")
+	tk.MustExec("insert t (id) values (3)")
+	tk.MustExec("update t set id = 4 where id = 1")
+	tk.MustExec("update t set id = 5 where id = 2")
+	tk.MustExec("update t set id = 6 where id = 3")
+	c.Assert(executor.GlobalMemoryUsageTracker.MaxConsumed(), Greater, int64(0))
+	c.Assert(executor.GlobalMemoryUsageTracker.BytesConsumed(), LessEqual, int64(0))
```
Comment (Yisaer) [inline executor/memory_test.go:137]: During Test I found that after finishing executing the dml statement, the `BytesConsumed` of the `GlobalMemoryTracker` is note 0.

After I dig into the source code, I fount that the un-zero result caused by the following code:

https://github.com/pingcap/tidb/blob/df6898d524134703f107154ca4d5166187b9d455/session/txn.go#L547-L549

https://github.com/pingcap/tidb/blob/df6898d524134703f107154ca4d5166187b9d455/session/txn.go#L569-L571

It seems the calculating logic for the memory usage between consuming and returning is different. Not sure whether it is a bug or not. WDYT? @qw4990

Comment (Yisaer) [reply executor/memory_test.go:137]: after update master branch, the test passed.

**Review by qw4990 — COMMENTED** [review]
~LGTM~

File: server/conn.go [hunk 2]
```diff
@@ -1316,6 +1316,18 @@ func (cc *clientConn) handleIndexAdvise(ctx context.Context, indexAdviseInfo *ex
 func (cc *clientConn) handleQuery(ctx context.Context, sql string) (err error) {
 	defer trace.StartRegion(ctx, "handleQuery").End()
 	sc := cc.ctx.GetSessionVars().StmtCtx
+	detachSessionTracker := func() {
```
Comment (SunRunAway) [inline server/conn.go:1319]: I do not think it's a good idea to handle this only on `handleQuery`, the other protocol like `handleStmtExecute` may be missed.

Comment (Yisaer) [reply server/conn.go]: good catch

**Pull request closed without merging**